mailpoet/lib/Subscription/Captcha/Validator/BuiltInCaptchaValidator.php

```diff
@@ -0,0 +1,142 @@
+<?php
+
+namespace MailPoet\Subscription\Captcha\Validator;
+
+use MailPoet\Models\Subscriber;
+use MailPoet\Subscribers\SubscriberIPsRepository;
+use MailPoet\Subscription\Captcha\CaptchaPhrase;
+use MailPoet\Subscription\Captcha\CaptchaSession;
+use MailPoet\Subscription\SubscriptionUrlFactory;
+use MailPoet\Util\Helpers;
+use MailPoet\WP\Functions as WPFunctions;
+
+class BuiltInCaptchaValidator implements CaptchaValidator {
+
+
+  /** @var SubscriptionUrlFactory  */
+  private $subscriptionUrlFactory;
+
+  /** @var CaptchaPhrase  */
+  private $captchaPhrase;
+
+  /** @var CaptchaSession  */
+  private $captchaSession;
+
+  /** @var WPFunctions  */
+  private $wp;
+
+  /** @var SubscriberIPsRepository  */
+  private $subscriberIPsRepository;
+
+  public function __construct(
+    SubscriptionUrlFactory $subscriptionUrlFactory,
+    CaptchaPhrase $captchaPhrase,
+    CaptchaSession $captchaSession,
+    WPFunctions $wp,
+    SubscriberIPsRepository $subscriberIPsRepository
+  ) {
+    $this->subscriptionUrlFactory = $subscriptionUrlFactory;
+    $this->captchaPhrase = $captchaPhrase;
+    $this->captchaSession = $captchaSession;
+    $this->wp = $wp;
+    $this->subscriberIPsRepository = $subscriberIPsRepository;
+  }
+
+  public function validate(array $data): bool {
+    $isBuiltinCaptchaRequired = $this->isRequired(isset($data['email']) ? $data['email'] : '');
+    if (!$isBuiltinCaptchaRequired) {
+      return true;
+    }
+    if (empty($data['captcha'])) {
+      throw new ValidationError(
+        $this->wp->__('Please fill in the CAPTCHA.', 'mailpoet'),
+        [
+          'redirect_url' => $this->subscriptionUrlFactory->getCaptchaUrl($this->getSessionId()),
+        ]
+      );
+    }
+    $captchaHash = $this->captchaPhrase->getPhrase();
+    if (empty($captchaHash)) {
+      throw new ValidationError(
+        $this->wp->__('Please regenerate the CAPTCHA.', 'mailpoet'),
+        [
+          'redirect_url' => $this->subscriptionUrlFactory->getCaptchaUrl($this->getSessionId()),
+        ]
+      );
+    }
+
+    if (!hash_equals(strtolower($data['captcha']), strtolower($captchaHash))) {
+      $this->captchaPhrase->resetPhrase();
+      throw new ValidationError(
+        $this->wp->__('The characters entered do not match with the previous CAPTCHA.', 'mailpoet'),
+        [
+          'refresh_captcha' => true,
+        ]
+      );
+    }
+
+    return true;
+
+  }
+
+  private function getSessionId() {
```

Comment:
I feel like this method belongs to `MailPoet\Subscription\Captcha:CaptchaSession` class, and the `init` method receives the `session id` and it has a logic to check if it needs to init or not, so this logic here is redundant

Reply:
Addressed in [db4fc58](https://github.com/mailpoet/mailpoet/pull/4196/commits/db4fc584c4e90875fb4269778007f022ca932608)